freqtrade/edge/__init__.py

```diff
@@ -0,0 +1,388 @@
+
+        self.fee = self.exchange.get_fee()
+
+    def calculate(self) -> bool:
+        pairs = self.config['exchange']['pair_whitelist']
+        heartbeat = self.config['edge']['process_throttle_secs']
+
+        if (self._last_updated is not None) and (
+                self._last_updated + heartbeat > arrow.utcnow().timestamp):
+            return False
+
+        data: Dict[str, Any] = {}
+        logger.info('Using stake_currency: %s ...', self.config['stake_currency'])
+        logger.info('Using stake_amount: %s ...', self.config['stake_amount'])
+        logger.info('Using local backtesting data (using whitelist in given config) ...')
+        # TODO: add "timerange" to Edge config
+        timerange = Arguments.parse_timerange(None if self.config.get(
+            'timerange') is None else str(self.config.get('timerange')))
+
+        data = optimize.load_data(
+            self.config['datadir'],
+            pairs=pairs,
+            ticker_interval=self.ticker_interval,
+            refresh_pairs=True,
+            exchange=self.exchange,
+            timerange=timerange
+        )
+
+        if not data:
+            logger.critical("No data found. Edge is stopped ...")
+            return False
+
+        preprocessed = self.tickerdata_to_dataframe(data)
+
+        # Print timeframe
+        min_date, max_date = self.get_timeframe(preprocessed)
+        logger.info(
+            'Measuring data from %s up to %s (%s days) ...',
+            min_date.isoformat(),
+            max_date.isoformat(),
+            (max_date - min_date).days
+        )
+        headers = ['date', 'buy', 'open', 'close', 'sell', 'high', 'low']
+
+        stoploss_range_min = float(self.edge_config.get('stoploss_range_min', -0.01))
+        stoploss_range_max = float(self.edge_config.get('stoploss_range_max', -0.05))
+        stoploss_range_step = float(self.edge_config.get('stoploss_range_step', -0.001))
+        stoploss_range = np.arange(stoploss_range_min, stoploss_range_max, stoploss_range_step)
+
+        trades: list = []
+        for pair, pair_data in preprocessed.items():
+            # Sorting dataframe by date and reset index
+            pair_data = pair_data.sort_values(by=['date'])
+            pair_data = pair_data.reset_index(drop=True)
+
+            ticker_data = self.advise_sell(
+                self.advise_buy(pair_data, {'pair': pair}), {'pair': pair})[headers].copy()
+
+            trades += self._find_trades_for_stoploss_range(ticker_data, pair, stoploss_range)
+
+        # Switch List of Trade Dicts (trades) to Dataframe
+        # Fill missing, calculable columns, profit, duration , abs etc.
+        trades_df = DataFrame(trades)
+
+        if len(trades_df) > 0:  # Only post process a frame if it has a record
+            trades_df = self._fill_calculable_fields(trades_df)
+        else:
+            trades_df = []
+            trades_df = DataFrame.from_records(trades_df, columns=BacktestResult._fields)
```

Comment:
is trades_df a list (`[]`) - or a dataframe ... ?
don't think both lines are necessary ...